### PR TITLE
Support integer weights in `hist` and remove dtype limitations

### LINKED ISSUE #3126: Cannot histogram integer count data
```Py
import scipp as sc
import numpy as np

N = 1000
counts = sc.ones(sizes={'event': N}, dtype=int, unit='counts')
tof = sc.array(dims=['event'], values=np.random.normal(size=N))
da = sc.DataArray(data=counts, coords={'tof': tof})
da.hist(tof=100)
```
fails with
Screenshot at 2023-04-21 11-54-11

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #3126.

As discussed in the linked issue, directly supporting all possible dtype combinations would require a very large number of template instantiations. With the change here I am converting to the common type of coord and edges, which avoid this. We thus have only `4*4+1*4 = 20` cases. Otherwise it would be `4*4*4+1*4 = 68`.

This adds `common_type`, which may be useful for other applications. Maybe it would even be worth exposing to Python?

Unrelated, I got a lot of compiler warnings about passing `std::string_view` by value, which apparently even made compilation fail with our current settings. I added references everywhere.

### CODE REVIEW

File: lib/dataset/test/histogram_test.cpp [hunk 1]
```diff
@@ -311,6 +336,34 @@ TEST_F(Histogram1DTest, coord_name_differs_dim) {
                                  Values{19, 12, 12}));
 }
 
+TEST_F(Histogram1DTest, int64_weights) {
```
Comment (nvaytet) [inline lib/dataset/test/histogram_test.cpp:339]: Also add a test for `int32`?

File: lib/dataset/histogram.cpp [hunk 2]
```diff
@@ -77,11 +78,21 @@ DataArray histogram(const DataArray &events, const Variable &binEdges) {
...
             return buckets::histogram(binned, binEdges_);
           }
-          return transform_subspan(events_.dtype(), dim,
-                                   binEdges_.dims()[dim] - 1,
-                                   subspan_view(cont_coord, event_dim_),
-                                   subspan_view(cont_data, event_dim_),
-                                   binEdges_, element::histogram, "histogram");
+          // Due to the combinatoric explosion of dtype combinations, we promote
+          // either the bin edges or the coord in case their dtypes mismatch.
+          // This is less efficient, but hopefully and edge case. If performance
```
Comment (nvaytet) [inline lib/dataset/histogram.cpp:83]:
```suggestion
          // This is less efficient, but hopefully an edge case. If performance
```

File: lib/dataset/test/histogram_test.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply lib/dataset/test/histogram_test.cpp:339]: What about the other 10+ cases? What would we gain?

See also discussion in #3251.

**Review by nvaytet — APPROVED** [review]
